doc/sphinx/addendum/sprop.rst

```diff
@@ -0,0 +1,197 @@
+
+Or more conveniently (but equivalently)
+
+.. coqtop::
+
+  Inductive Squash (A:Type) : SProp := squash : A -> Squash A.
+
+Most inductives types defined in :math:`\SProp` are squashed types,
+i.e. they can only be eliminated to construct proofs of other strict
+propositions. The exceptions are empty types and unit types.
+
+.. coqtop::
+
+   Inductive sEmpty : SProp := .
+   Check sEmpty_rect.
+
+   Inductive sUnit : SProp := stt.
+   Check sUnit_rect.
+
+Here unit types means inductive types with no indices and one
+constructor with no arguments, so for instance the identity is
+squashed:
+
+.. coqtop::
+
+   Fail Check fun A (a b:A) (p:seq a b) =>
+     match p in seq _ x return eq x a with srefl => eq_refl end.
+
+Primitive records in :math:`\SProp` are allowed when fields are strict
+propositions, for instance:
+
+.. coqtop::
+
+   Set Primitive Projections.
+   Record sProd (A B : SProp) : SProp := { sfst : A; ssnd : B }.
+
+On the other hand, to avoid having definitionally irrelevant types in
+non-:math:`\SProp` sorts (through record η-extensionality), primitive
+records in relevant sorts must have at least one relevant field.
+
+.. coqtop::
+
+   Set Warnings "+non-primitive-record".
+   Fail Record rBox (A:SProp) : Prop := rbox { runbox : A }.
+
+   Record ssig (A:Type) (P:A -> SProp) : Type := { spr1 : A; spr2 : P spr1 }.
+
+Note that ``rBox`` works as an emulated record, which is equivalent to
+the Box inductive.
+
+Encodings for strict propositions
+---------------------------------
+
+By using empty and unit types as base values, we can encode other
+strict propositions. For instance:
+
+.. coqtop::
+
+  Definition is_true (b:bool) : SProp := if b then sUnit else sEmpty.
+
+  Definition is_true_eq_true b : is_true b -> true = b
+    := match b with
+       | true => fun _ => eq_refl
+       | false => sEmpty_ind _
+       end.
+
+  Definition eq_true_is_true b (H:true=b) : is_true b
+    := match H in _ = x return is_true x with eq_refl => stt end.
+
+Issues with non-cumulativity
+----------------------------
+
+During normal term elaboration, we don't always know that a type is a
+strict proposition early enough. For instance:
+
+.. coqtop::
+
+   Definition constant_0 : ?[T] -> nat := fun _ : sUnit => 0.
+
+While checking the type of the constant, we only know that ``?[T]``
+must inhabit some sort. Putting it in some floating universe ``u``
+would disallow instantiating it by ``sUnit : SProp``.
+
+In order to make the system usable without having to annotate every
+instance of :math:`\SProp`, we consider :math:`\SProp` to be a subtype
+of every universe during elaboration (i.e. outside the kernel). Then
+once we have a fully elaborated term it is sent to the kernel which
+will check that we didn't actually need cumulativity of :math:`\SProp`
+(in the example above, ``u`` doesn't appear in the final term).
+
+This means that some errors will be delayed until ``Qed``:
```

Comment:
There is no equivalent of `Guarded` then to use the kernel in the middle of a proof?

Reply:
>There is no equivalent of Guarded then to use the kernel in the middle of a proof?

No. Maybe someday ^^

Reply:
But wouldn't it be easy to add it?

Reply:
>But wouldn't it be easy to add it?

Not so easy, or at least I don't see it. PRs welcome ;)